:hammer: Updating state.ts to deprecate existing precondition apis and updating them with new

Saurabhpatil-dev opened the pull request:
Hi @mitschabaude ,
As discussed in issue #1247 , I have replaced old api by deprecating them rather than removing.  
Question:
> Do we need to update this `assertStatePrecondition` too ?

Please review and let know if any changes are required.

mitschabaude commented on src/lib/state.ts:
```diff
@@ -203,6 +215,23 @@ function createState<T>(): InternalStateType<T> {
+      this._contract.wasConstrained = true;
+    },
+
     assertEquals(state: T) {
       if (this._contract === undefined)
```
It would be nice if you could get rid of the code duplication, by making the `assert*` methods just call the `require*` versions. Something like
```ts
assertEquals(state: T) {
   return this.requireEquals(state);
}
```
etc

Saurabhpatil-dev replied:
Hi @mitschabaude ,

As discussed above I have removed the duplicate code by making assert* call the require* version. 
As well as build the project locally and adding screenshots of it for your reference.
![State_ts_deprecated](https://github.com/o1-labs/o1js/assets/55076843/636f0899-a466-4fec-bc32-b12b8e2421e8)
![State_ts_deprecated_01](https://github.com/o1-labs/o1js/assets/55076843/5792b9dd-1e89-4861-87fd-192ee7145558)


Also, I am making changes to CHANGELOG.md for first time so can you please help to review it. I have added change in Unreleased for now. Let me know if we need move it.

Please let me know if any change is required.

Thank You 😃

mitschabaude reviewed (commented):
Very nice! If you're up for it, two things are needed as a follow-up (feel free to do this in a later PR)
* update `/src/examples` - maybe grep for `.getAndAssertEquals` etc, check if it's now deprecated and replace with require version
* add an entry to `CHANGELOG.md` (under `## Changed`) about this change

Saurabhpatil-dev commented:
Hi @mitschabaude ,

As discussed above I have removed the duplicate code by making assert* call the require* version.
As well as built the project locally and adding screenshots of it for your reference.

![State_ts_deprecated](https://github.com/o1-labs/o1js/assets/55076843/636f0899-a466-4fec-bc32-b12b8e2421e8)
![State_ts_deprecated_01](https://github.com/o1-labs/o1js/assets/55076843/5792b9dd-1e89-4861-87fd-192ee7145558)


Also, I am making changes to CHANGELOG.md for first time so can you please help to review it. I have added change in Unreleased for now. Let me know if we need move it.

Please let me know if any change is required.

Thank You 😃

mitschabaude commented on CHANGELOG.md:
```diff
@@ -19,7 +19,13 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
-> No unreleased changes yet
+### Changed
+
+- Preconditioned asset* in state have been renamed to require* :
+  For example:
+  - `this.x.getAndAssertEquals()` is now `this.x.getAndRequireEquals()` https://github.com/o1-labs/o1js/pull/1263
+  - `this.x.assertEquals(x)` is now `this.x.requireEquals(x)` https://github.com/o1-labs/o1js/pull/1263
+  - `this.x.assertNothing()` is now `this.x.requireNothing()` https://github.com/o1-labs/o1js/pull/1263
```
```suggestion
- Change precondition APIs to use "require" instead of "assert" as the verb, to distinguish them from provable assertions.
  - `this.x.getAndAssertEquals()` is now `this.x.getAndRequireEquals()` https://github.com/o1-labs/o1js/pull/1263
  - `this.x.assertEquals(x)` is now `this.x.requireEquals(x)` https://github.com/o1-labs/o1js/pull/1263
  - `this.x.assertNothing()` is now `this.x.requireNothing()` https://github.com/o1-labs/o1js/pull/1263
```
